### PR TITLE
Added collecting exec_locals

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

File: protostar/starknet/cheatcode.py [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 
 class Cheatcode(BusinessLogicSysCallHandler, HintLocal):
+    _exec_locals = {}
+
```
Comment (kasperski95) [inline protostar/starknet/cheatcode.py:24]: - Why is this a static property of Cheatcode? Only `ReflectCheatcode` uses it.
- This property uses private property notation, but it's used as a public property.
- A comment explaining this property would be useful.

File: protostar/starknet/cheatable_execute_entry_point.py [hunk 2]
```diff
@@ -73,7 +75,9 @@ def _run(
 
         # Run the specified contract entry point with given calldata.
         with wrap_with_stark_exception(code=StarknetErrorCode.SECURITY_ERROR):
-            runner = CairoFunctionRunner(program=contract_class.program, layout="all")
+            runner = CheatableCairoFunctionRunner(  # <-- MODIFICATION # TODO
+                program=contract_class.program, layout="all"
+            )
```
Comment (kasperski95) [inline protostar/starknet/cheatable_execute_entry_point.py:80]: TODO?

File: protostar/starknet/cheatable_cairo_vm.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            exec_locals["current_step"] = self.current_step
+            exec_locals["ids"] = hint.consts(pc, ap, fp, memory)
+
+            exec_locals["vm_load_program"] = self.load_program
+            exec_locals["vm_enter_scope"] = self.enter_scope
+            exec_locals["vm_exit_scope"] = self.exit_scope
+            exec_locals.update(self.static_locals)
+            exec_locals.update(self.builtin_runners)
+
+            # --- MODIFICATIONS START --- # TODO
```
Comment (kasperski95) [inline protostar/starknet/cheatable_cairo_vm.py:25]: TODO?

File: protostar/starknet/cheatable_cairo_function_runner.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+from typing import Optional, Dict, Any
+from starkware.cairo.common.cairo_function_runner import CairoFunctionRunner
+from protostar.starknet.cheatable_cairo_vm import CheatableVirtualMachine
+
+
+class CheatableCairoFunctionRunner(CairoFunctionRunner):
+    # MODIFICATION vm_class=VirutalMachine -> vm_class=CheatableVirtualMachine # TODO
```
Comment (kasperski95) [inline protostar/starknet/cheatable_cairo_function_runner.py:7]: You can describe the class using triple double-quotes (see: ForkableStarknet for a reference)
`
"""
"""
`

File: protostar/commands/test/cheatcodes/reflect_cheatcode.py [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
+
+class ReflectClass:
+    def __getattr__(self, name):
+        return getattr(Reflector(Cheatcode._exec_locals["ids"]), name)
 
     # pylint: disable=R0201
-    def reflect(self, ids: VmConsts) -> Reflector:
-        return Reflector(ids)
+    def get(self):
+        raise CheatcodeException("reflect", "get() called on no value.")
```
Comment (kasperski95) [inline protostar/commands/test/cheatcodes/reflect_cheatcode.py:21]: Why is this needed?

File: protostar/starknet/cheatcode.py (hunk 1 above)
Comment (Radinyn) [reply protostar/starknet/cheatcode.py:24]: 1. It allows all future cheatcodes to use execution locals and they can be very useful.
2, 3. Good points 👍

File: protostar/starknet/cheatable_execute_entry_point.py (hunk 2 above)
Comment (Radinyn) [reply protostar/starknet/cheatable_execute_entry_point.py:80]: I've seen we have used this convention in other cheatables, we can change it though.

File: protostar/starknet/cheatcode.py (hunk 1 above)
Comment (kasperski95) [reply protostar/starknet/cheatcode.py]: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

File: protostar/starknet/cheatable_execute_entry_point.py (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply protostar/starknet/cheatable_execute_entry_point.py]: This TODO is something is my leftover, i forgot to removed looong time ago. Sorry about that, this TODO should be removed 😅

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
We should avoid using global state, ase we discused privately

File: protostar/commands/test/cheatcodes/reflect_cheatcode.py [hunk 6]
```diff
@@ -1,17 +1,12 @@
...
     def name(self) -> str:
         return "reflect"
 
-    def build(self) -> Callable[[VmConsts], Reflector]:
-        return self.reflect
-
-    # pylint: disable=R0201
-    def reflect(self, ids: VmConsts) -> Reflector:
-        return Reflector(ids)
+    def build(self) -> DelayedBuilder:
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/test/cheatcodes/reflect_cheatcode.py:11]: Please change build -> Any for Cheatcode type

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Very nice, some minor comments

File: protostar/starknet/cheatable_cairo_vm.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            exec_locals["vm_exit_scope"] = self.exit_scope
+            exec_locals.update(self.static_locals)
+            exec_locals.update(self.builtin_runners)
+
+            # --- MODIFICATIONS START ---
+            for name, value in exec_locals.items():
+                if isinstance(value, DelayedBuilder):
+                    exec_locals[name] = value.internal_build(exec_locals)
+
+            self.exec_hint(hint.compiled, exec_locals, hint_index=hint_index)
```
[the remark this answers is not in the document]
Comment (Radinyn) [reply protostar/starknet/cheatable_cairo_vm.py:34]: Do we delete other cheatcodes? It's exactly the same situation here.

**Review by MaksymilianDemitraszek — APPROVED** [review]
